File: mobile/lib/chartgeom.js
// chartgeom.js: pure geometry for the trend line charts. Maps a series of
// values onto x,y pixel positions inside a box, so the Skia component stays a
// dumb renderer and this math can be unit tested. No Skia, no React here.

const num = (x) => (Number.isFinite(Number(x)) ? Number(x) : 0);

// sharedMax(seriesList) -> the max value across several series, so two lines
// on one chart share one honest scale (income and spending must be comparable
// on sight, never independently normalized).
export function sharedMax(seriesList) {
  const all = (Array.isArray(seriesList) ? seriesList : [])
    .flatMap((s) => (Array.isArray(s) ? s.map(num) : []));
  return all.length ? Math.max(...all, 0) : 0;
}

// linePointsScaled(values, maxValue, width, height, pad) -> [{x, y}] for a
// polyline, scaled against a caller-supplied max (use sharedMax for
// multi-series charts so every line shares one scale).
//   values  the series, oldest first. Non-numeric entries count as 0.
//   width   drawable width in px
//   height  drawable height in px
//   pad     inset from every edge so dots at the extremes are not clipped
// Scale: x spreads points evenly from left pad to right pad. y runs from 0 at
// the bottom to maxValue at the top, with 8 percent headroom so the top dot
// never kisses the edge. An all-zero series draws along the baseline. A single
// point lands centered horizontally.
export function linePointsScaled(values, maxValue, width, height, pad = 8) {
  const list = Array.isArray(values) ? values.map(num) : [];
  if (list.length === 0) return [];
  const w = Math.max(num(width), 2 * pad + 1);
  const h = Math.max(num(height), 2 * pad + 1);
  const innerW = w - 2 * pad;
  const innerH = h - 2 * pad;
  const scaled = Math.max(0, num(maxValue)) * 1.08;
  const denom = scaled > 0 ? scaled : 1;
  const n = list.length;
  return list.map((v, i) => ({
    x: n === 1 ? w / 2 : pad + (innerW * i) / (n - 1),
    y: pad + innerH * (1 - Math.max(0, v) / denom),
  }));
}

// linePoints(values, width, height, pad) -> single-series convenience: scaled
// against the series' own max.
export function linePoints(values, width, height, pad = 8) {
  const list = Array.isArray(values) ? values.map(num) : [];
  return linePointsScaled(list, list.length ? Math.max(...list, 0) : 0, width, height, pad);
}

File: flutter/lib/money/chartgeom.dart
// Pure geometry for the trend line charts, ported 1:1 from
// mobile/lib/chartgeom.js. Maps series values onto x,y positions inside a
// box so the chart widget stays a dumb renderer and this math is golden
// verified against the RN module. No Flutter imports here on purpose.

import 'ledger.dart' show amountOf;

/// The max across several series, so multiple lines on one chart share one
/// honest scale (income and spending must be comparable on sight).
double sharedMax(dynamic seriesList) {
  final all = <double>[];
  for (final s in (seriesList is List ? seriesList : const [])) {
    if (s is List) {
      for (final v in s) {
        all.add(amountOf(v));
      }
    }
  }
  if (all.isEmpty) return 0;
  var max = 0.0;
  for (final v in all) {
    if (v > max) max = v;
  }
  return max;
}

/// Points for a polyline, scaled against a caller-supplied max. x spreads
/// evenly between the pads; y runs 0 at the bottom to maxValue at the top
/// with 8 percent headroom. All-zero draws along the baseline; a single
/// point lands centered horizontally.
List<Map<String, double>> linePointsScaled(
    dynamic values, dynamic maxValue, dynamic width, dynamic height,
    [double pad = 8]) {
  final list = [
    for (final v in (values is List ? values : const [])) amountOf(v),
  ];
  if (list.isEmpty) return [];
  final minSize = 2 * pad + 1;
  final w = amountOf(width) > minSize ? amountOf(width) : minSize;
  final h = amountOf(height) > minSize ? amountOf(height) : minSize;
  final innerW = w - 2 * pad;
  final innerH = h - 2 * pad;
  final mv = amountOf(maxValue);
  final scaled = (mv > 0 ? mv : 0) * 1.08;
  final denom = scaled > 0 ? scaled : 1;
  final n = list.length;
  return [
    for (var i = 0; i < n; i++)
      {
        'x': n == 1 ? w / 2 : pad + (innerW * i) / (n - 1),
        'y': pad +
            innerH * (1 - (list[i] > 0 ? list[i] : 0) / denom),
      },
  ];
}

/// Single-series convenience: scaled against the series' own max.
List<Map<String, double>> linePoints(
    dynamic values, dynamic width, dynamic height,
    [double pad = 8]) {
  final list = [
    for (final v in (values is List ? values : const [])) amountOf(v),
  ];
  var max = 0.0;
  for (final v in list) {
    if (v > max) max = v;
  }
  return linePointsScaled(list, list.isEmpty ? 0 : max, width, height, pad);
}
